students/pkoleyni/Lesson05/mailroom_3.py

```diff
@@ -0,0 +1,186 @@
+                Then it will add name to the donor_list_dic dictionary
+
+    """
+
+    if name not in donor_list_dic:
+        print('\"{}\" is not in the list of the donors, Adding to the list of donors'.format(name))
+        donor_list_dic[name] = {'donation': []}
+    else:
+        print('{} is already in the list of donors'.format(name))
+
+
+def donation_list(name):
+    """
+    :param name: name of a donor
+    :return: list of all the donation made by a donor
+    """
+    return donor_list_dic[name]['donation']
+
+
+def add_donation(name, donation):
+    """
+    :param name: name of a donor
+    :param donation: amount of donation made
+    :return:
+    """
+    donation_list(name).append(donation)
+    update_total_donation_to_dic(name)
+
+
+def update_total_donation_to_dic(name):
+    total_given = sum(donation_list(name))
+    donor_list_dic[name]['total_donation'] = total_given
+
+
+def thank_you():
+    """
+    This function is taking care of the required actions when user select thank you from the menu
+    :return:
+    """
+    usr_name_input = input('Enter full name of a donor or enter \"List\" for complete list of doors > ')
+    while usr_name_input == 'List':
+        [print(donor) for donor in donor_list_dic]
+        usr_name_input = input('Enter full name of a donor or enter \"List\" for complete list of doors > ')
+    check_add_name(usr_name_input)
+    usr_donation_input = input('Enter a donation amount > ')
+    try:
+        usr_donation_input = int(usr_donation_input)
+    except ValueError:
+        print('ValueError: {} is a {}, Enter a digit for donation value'.format(usr_donation_input,type(usr_donation_input)))
+    else:
+        add_donation(usr_name_input, usr_donation_input)
+        thank_you_message(usr_name_input, usr_donation_input)
+
+
+def thank_you_message(name, donation):
+    """
+    :param name: name of a donor to send thank you letter
+    :param donation: Amount of donation
+    :return:
+    """
+    print()
+    print("Thank you letter for {}".format(name))
+    print('=' * 50)
+    print("Dear {}".format(name))
+    print("We received your {} donation".format(donation))
+    print("Thank you for your donation\n")
+
+
+def sort_key(key):
+    """
+
+    :param key: Key defines the item in the list to be used for sorting
+    :return:
+    """
+    return key[1]
+
+
+def report():
+    """
+
+    :return: This function created new list with the items required in the report, sort the list and print the report
+    """
+
+
+    donor_names = [donor for donor in donor_list_dic]
+    total_given = [sum(donation_list(donor)) for donor in donor_names]
+    num_of_gifts = [len(donation_list(donor)) for donor in donor_names]
+    average_gifts = [t/g for t in total_given for g in num_of_gifts]
+
+    report_list = []
+    for i in range(len(donor_names)):
+        report_list.append([donor_names[i], total_given[i], num_of_gifts[i], average_gifts[i]])
```

Comment:
including this as well, you can combine into a single loop and get rid of loops in lines 101 - 104